src/compiler/mod.rs

```diff
@@ -0,0 +1,65 @@
+//! The compiler for `ronkathon` circuits.
+
+// TODO: Remove this once the module is fleshed out.
+#![allow(missing_docs)]
+use petgraph::graph::{DiGraph, NodeIndex};
+
+// TODO: Goal, allow someone to use this library to write rust to generate arithmetic circuits
+// basically.
+// ```
+// use ronkathon::compiler::*;
+//
+// fn main() {
+//   let a: Input = Variable::public(7);
+//   let b: Input = Variable::private(3);
+//   let b: Expression = a * c;
+// }
+// ```
+// So we can do things like `impl Add`, `impl Mul` for variables and make them into gates?
+
+// Above seems done. Now we need to have a way to unravel a collection of expressions into a
+// circuit that may have the same inputs and outputs as the expressions. Inputs are going to be
+// the terminal variables found by fully unravelling expressions and they should be named. The
+// fully ravelled expressions are the outputs, and they can also be named
+use super::*;
+
+pub mod dsl;
+
+pub struct Wire {
+  pub input:  Connection,
+  pub output: Connection,
+}
+
+pub enum Connection {
+  Input(Publicity),
+  Output,
+  Internal,
+}
+
+pub enum Publicity {
+  Public(u32),
+  Private(u32),
+}
+
+pub enum Operation {
+  Add,
+  Mul,
+  Sub,
+}
+
+// TODO: Want to make it so gates own a reference for who they are connected to, and circuits own
+// the gates. This way gates cannot be improperly connected to other gates.
+pub struct Gate {
+  pub left:   Wire,
+  pub right:  Wire,
+  pub output: Wire,
+  pub op:     Operation,
+}
+
+pub struct Circuit {
+  pub gates: DiGraph<Gate, Wire>,
```

Comment:
maybe instead of picturing a circuit as graph, a simple tuple of (Gate, Constraint) that extends to rows and column in execution trace?

Reply:
I think I do want to trash the graph, this is a relic of my first pass. I will remove it for now. 

One thing I also want to do is impl a nice Display for our circuits, so that's on my mind as far as this goes.